runtime/src/iree/hal/drivers/hip/hip_device_group_device.c

```diff
@@ -0,0 +1,1535 @@
+    }
+
+    // Create the default dispatch stream for the device.
+    if (iree_status_is_ok(status)) {
+      status = IREE_HIP_RESULT_TO_STATUS(
+          symbols, hipStreamCreateWithFlags(
+                       &device->device_contexts[i].hip_dispatch_stream,
+                       hipStreamNonBlocking));
+    }
+
+    if (iree_status_is_ok(status)) {
+      for (uint32_t j = 0; j < device_count && iree_status_is_ok(status); ++j) {
+        if (i == j) {
+          continue;
+        }
+        status = IREE_HIP_RESULT_TO_STATUS(
+            symbols, hipDeviceEnablePeerAccess(devices[j], 0));
+      }
+    }
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_device_group_device_initialize_internal(
+        driver, identifier, params, device, symbols, nccl_symbols,
+        host_allocator);
+  } else {
+    for (uint32_t i = 0; i < device_count && iree_status_is_ok(status); ++i) {
+      if (device->device_contexts[i].hip_dispatch_stream)
+        symbols->hipStreamDestroy(
+            device->device_contexts[i].hip_dispatch_stream);
+      // NOTE: This function return hipSuccess though doesn't release the
+      // primaryCtx by design on HIP/HCC path.
+      if (device->device_contexts[i].hip_context)
+        symbols->hipDevicePrimaryCtxRelease(
+            device->device_contexts[i].hip_device);
+    }
+  }
+
+  iree_event_pool_t* host_event_pool = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_event_pool_allocate(params->event_pool_capacity,
+                                      host_allocator, &host_event_pool);
+  }
+
+  for (uint32_t i = 0; i < device_count && iree_status_is_ok(status); ++i) {
+    if (iree_status_is_ok(status)) {
+      status = iree_hal_hip_event_pool_allocate(
+          symbols, params->event_pool_capacity, host_allocator,
+          device->device_contexts[i].hip_context,
+          &device->device_contexts[i].device_event_pool);
+    }
+  }
+
+  iree_hal_hip_timepoint_pool_t* timepoint_pool = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_timepoint_pool_allocate(
+        host_event_pool, device_count, device->device_contexts,
+        params->event_pool_capacity, host_allocator, &timepoint_pool);
+  }
+
+  if (iree_status_is_ok(status)) {
+    device->host_event_pool = host_event_pool;
+    device->timepoint_pool = timepoint_pool;
+  } else {
+    // Release resources we have accquired after HAL device creation.
+    if (timepoint_pool) iree_hal_hip_timepoint_pool_free(timepoint_pool);
+    for (uint32_t i = 0; i < device_count; ++i) {
+      if (device->device_contexts[i].device_event_pool)
+        iree_hal_hip_event_pool_release(
+            device->device_contexts[i].device_event_pool);
+    }
+    if (host_event_pool) iree_event_pool_free(host_event_pool);
+    // Release other resources via the HAL device.
+    iree_hal_device_release((iree_hal_device_t*)device);
+    device = NULL;
+  }
+  *out_device = (iree_hal_device_t*)device;
```

Comment:
only do this in the if(ok) - the else doesn't need to set it - having the device = NULL and knowing that device may be null here is extra cognitive load and something that's easier to miss (it's 2 lines instead of 1 and non-local to the condition - a reader has to check the condition, then skip down to see what happens, then skip back up to the else to see how the state changes, etc).